### ISSUE TITLE
Consider using end pattern as bail out for embedded languages

### ISSUE THREAD

**Issue by alexr00** [microsoft/vscode-textmate#207]
This started as a discussion in https://github.com/jlelong/vscode-latex-basics/issues/58.

The problem is that languages that are embedded within other languages have a high potential to not work as expected. When there's a problem in the embedded grammar, then it can easily escape into the primary grammar. 

Key comment from @jlelong copied below:

> @jlelong would this be something that you would define in your textmate grammar at the point that you embed the grammar?

@alexr00 yes exactly. For instance, embedding the Python language typically looks like

```json
{
	"begin": "some_begin_pattern",
	"end": "some_end_pattern",
	"contentName": "source.python",
	"patterns": [
		{
			"include": "source.python"
		}
	]
}
```

It would be nice if the top level grammar could escape from the Python one when the end pattern is reached. I originally thought a new pattern was needed to trigger the bail out but I believe it will always be the same as the end pattern.

Currently, we provide a modified C++ grammar which includes the bail out pattern. This approach would make it useless. Embedding a grammar would much more robust.

**Comment by alexdima** [comment]
I share the sentiment, I find that this is one of the fundamental flaws of TextMate. When including another grammar, there should be a straight-forward way to specify a bail out end pattern. 

The fundamental problem here is that we want to be good open source citizens and we should not deviate from TM's implementation to avoid fracturing the TM grammar comunity. However, if we do decide to add this, we should try to add it in a minimal invasive way, perhaps as an optional property when including another grammar.

cc @hediet

**Comment by jeff-hykin** [comment]
I support your thoughts exactly.

I think there is a way forward, but I could use some help.

As far as I know, Textmate isn't a specification. There's no central comprehensive documentation, there's a nearly complete lack of standards when it comes to textmate-scoping names, and there's no working group for resolving design flaws or specifying "correct" behavior. I documented some of the obscure features on my own, but lack anywhere meaninful to put it. I've also had many conversations with grammar maintainers and old Atom contributors about standardizing textmate scopes, but we also lacked a central place to have a conversation or publish any agreed-upon result.

I think a central spec-repo (similar to [WebAssembly's](https://github.com/WebAssembly/spec)) could make it where a realatively small feature like bailouts could go from bad-fracturing-change to a helpful-new-addition that other textmate engines could agree upon and take advantage of.

I can do a lot of the work towards that, bringing the feature documentation, formalizing existing pseudo-standards, contacting contributors/maintainers of major Textmate implementations, but I would need the backing/support of vscode-textmate to add/start with some form of legitmacy.

**Comment by hediet** [comment]
@jeff-hykin I think such effort would be beneficial. Please note that we don't have resources at the moment to assist with that.

Also note that there are [subtle bugs](https://github.com/microsoft/vscode-textmate/issues/174) in our implementation.

I'd rather try to bring tree sitter tokenization support to VS Code than spending a significant amount of energy on improving textmate support.

**Comment by jeff-hykin** [comment]
> I'd rather try to bring tree sitter tokenization support to VS Code

Woah! As per https://github.com/microsoft/vscode/issues/50140 I thought tree-sitter support was a non-starter. If VS Code is even considering adding support, thats huge news to me. I'd be happy for all effort to be spent towards tree sitter support.

**Comment by RedCMD** [comment]
I thought TextMate was just a spec
but seems to also be a program?
idk if this is the official documentation
v1: https://macromates.com/manual/en/
v2: https://macromates.com/textmate/manual/

it has its opinions about scope names. which vscode dark seems to follow exactly?
https://macromates.com/manual/en/language_grammars#naming-conventions

> I documented some of the obscure features on my own

@jeff-hykin may I see that please?


`while` - `end` rules are more strict than `begin` - `end`
but also much more limited
prob not useable

I don't think injections can be used to solve this problem either
as they are only good for expanding rather than halting

**Comment by jeff-hykin** [comment]
Yeah I don't consider [https://macromates.com/](https://macromates.com/manual/en/) "comprehensive" by any means I don't even think it mentions `while` or `applyEndPatternLast`. The convention guide fails for all but the most basic/sterotypical scopes, and like you say it's closer to a personal opinion rather than a standard. The only textmate spec I'm aware of is behavior comparisons against the original implementation.

This could use updating/enhancing, but here's an example of some documention for `while` https://github.com/jeff-hykin/better-cpp-syntax/blob/fe873cdfacd1df7072e7b8c95df3df369c1ffcaa/documentation/library/textmate_while.md
I think I have one somewhere for `applyEndPatternLast`, and I've got several unfinished "How to use Textmate (without wanting to jump off a bridge)" drafts that would-be/are comprehensive.

**Comment by zm-cttae** [comment]
I believe these were new features meant to be documented in Textmate 2 but Apple didn't dedicate any resources to that yet.

This is all we have and will have for a long time:

https://macromates.com/manual/en/language_grammars#example_grammar

**Comment by zm-cttae** [comment]
Why isn't this being done by a tracking/hoisted variable that stores the end pattern when embedding begins?

**Comment by RedCMD** [comment]
@zm-cttae it does already
the problem is that the inner embedded language can push the end pattern outwards 
which is a feature, not a bug
tho not being able to disable it is the issue
but then if that setting did exist, it could cause just as many problems as it fixes
as it can be quite hard to deduce if the end pattern is inside a string etc without knowing the context around it

**Comment by zm-cttae** [comment]
It should be possible to filter the setter for that end pattern hoisted variable to a `begin` for embedded languages only.

**Comment by zm-cttae** [comment]
Another relevant technique: push a "stack" array of patterns for end conditions, alongside whether they end a language embed.

**Comment by RedCMD** [comment]
VSCode wants to keep parity with Apple's TextMate 2.0
so this will never be fixed

currently theres 3 options
1. modify the embedded grammar
   * requires a local copy of the grammar
   * and knowledge of how the grammar works
2. use `begin`/`while`
   *  `while` continues at the line after `begin` and keeps going down each line as long as it matches
   *  `while` will stop on the line that doesn't match
   *  `while` won't let embedded grammars escape
      however this is actually a vscode-textmate bug
      TextMate 2.0 and Github do allow embedded rules to escape out
3. use a capture
   * `captures` phyiscally capture the text and won't allow embedded grammars to escape
   * not multi-line
   * can be performance heavy on very very long lines (like 50-100k+ long lines)
   * pair with a multi-line fallback

<img width="656" height="348" alt="Image" src="https://github.com/user-attachments/assets/ff8bbb22-5b8d-41fa-bb61-d762ed32fc24" />

```json textmate
"repository": {
	"string_single-line": {
		"match": "(\")((?>[^\\\\\"]++|\\\\.))(\")",
		"captures": {
			"0": { "name": "string.quoted.double.languageId" },
			"1": { "name": "punctuation.definition.string.begin.languageId" },
			"2": {
				"patterns": [
					{
						"match": "\\\\.",
						"name": "constant.character.escape.languageId"
					},
					{ "include": "source.c" }
				]
			},
			"3": { "name": "punctuation.definition.string.end.languageId" }
		}
	},
	"string_multi-line": {
		"begin": "\"",
		"end": "\"",
		"beginCaptures": { "0": { "name": "punctuation.definition.string.begin.languageId" } },
		"endCaptures": { "0": { "name": "punctuation.definition.string.end.languageId" } },
		"name": "string.quoted.double.languageId",
		"patterns": [
			{
				"match": "\\\\.",
				"name": "constant.character.escape.languageId"
			}
		]
	}
}
```

**Comment by bentsherman** [comment]
(My agent hit this issue while adding support for embedding highlighting in Nextflow scripts.)

A concrete first-party repro, since the discussion above is mostly abstract: VS Code's own markdown grammar visibly mis-highlights shell in a fenced block.

````markdown
```bash
fastqc \
    --threads 4 \
    --memory 8 \
    out
```
````

Tokenized with `vscode-textmate` + `markdown.tmLanguage.json` + `shell-unix-bash.tmLanguage.json`:

```
"fastqc"             entity.name.command.shell
"    --threads 4 \"  meta.statement.command.name.shell   <- whole line swallowed
"-memory"            constant.other.option
"    out"            meta.statement.command.name.shell   <- swallowed
```

Alternating lines lose their highlighting. The same text in a `.sh` file is correct.

Cause is the `while`-based embedding markdown uses to avoid exactly the swallowing problem this issue is about. `\G` resolves to the end of the last stacked `while`, so the region re-anchors it at column 0 on every line, and shellscript's `command_name_range` (`"begin": "\\G"`) re-opens mid-command. Spec-conformant, per the quote in #237.

So the two options for embedding are currently: `end`, and the embedded grammar escapes into the host; or `while`, and `\G`-anchored rules in the embedded grammar misfire on every line. A bail-out pattern on the include would remove the need to choose. Reporting it here rather than opening a duplicate.
